client/src/components/room/streamGate.tsx

```diff
@@ -0,0 +1,136 @@
+import {
+  Call,
+  StreamCall,
+  StreamVideo,
+  StreamVideoClient,
+  User,
+} from "@stream-io/video-react-sdk";
+
+import { useCallback, useEffect, useState } from "react";
+
+import { logError, logInfo } from "../../lib/logger";
+import { getStreamAdminCredentials } from "../../lib/server-api";
+import { useAdminStore } from "../../stores/adminStore";
+import { useRoomStore } from "../../stores/roomStore";
+
+const publicUser: User = { type: "anonymous" };
+
+/**
+ * StreamGate provides context for being in a Stream call as an admin or viewer. Upon initializing, it will:
+ *
+ * - (As an admin) Get or create a stream call and admin RTMPS credentials
+ * - (All roles) Join the call as an admin or anonymous viewer
+ *
+ * Upon leaving the page, the user will automatically leave the call.
+ *  */
+const StreamGate: React.FunctionComponent<{
+  children: any;
+  roomID: string;
+  streamPlaybackID?: string;
+  anonymousOnly: boolean;
+}> = ({ children, roomID, streamPlaybackID, anonymousOnly }) => {
+  const isAdmin = useAdminStore(useCallback((s) => s.isAdmin, []));
+
+  const [state, setState] = useState<{
+    client: StreamVideoClient;
+    call: Call;
+    rtmpsDetails: RtmpsDetails | null;
+  }>();
+
+  const useAdmin = !anonymousOnly && isAdmin;
+
+  useEffect(() => {
+    getClient(roomID, useAdmin).then(([myClient, rtmpsDetails]) => {
+      if (!streamPlaybackID) {
+        // Technically, streamPlaybackID may be indirectly set in the middle of getClient()
+        // so there is a race condition here. If this happens, we will call getClient twice,
+        // and that's okay.
+        return;
+      }
+
+      const myCall = myClient.call("livestream", streamPlaybackID);
+
+      myCall
+        .get()
+        .then(() => {
+          logInfo("Setting state for call ID " + streamPlaybackID);
+          setState({ client: myClient, call: myCall, rtmpsDetails });
+        })
+        .catch((e) => logError("Failed to retrieve call details", e));
+    });
+  }, [roomID, streamPlaybackID, useAdmin]);
+
+  useEffect(() => {
+    if (!state) {
+      return;
+    }
+
+    logInfo("....JOINING CALL....");
+
+    // Ensure camera is disabled by default on page load.
+    state.call.camera.disable();
+
+    state.call
+      .join()
+      .then(() => {
+        logInfo("Joined call from StreamGate");
+      })
+      .catch((e) => {
+        logError("Failed to join call", e);
+      });
+
+    return () => {
+      state.call.leave().catch((e) => {
```

Comment:
Wondering if this may leave the call when the tab is backgrounded? Or just when closed?

Reply:
From a quick googling, the cleanup hooks only run when the tabs are closed, not backgrounded. This seems to match what I remember when I was writing this PR and testing the changes, although that was a little while ago.